src/main/java/de/medizininformatik_initiative/process/data_sharing/service/merge/DecryptValidateAndInsertDataSet.java

```diff
@@ -0,0 +1,437 @@
+			Resource resource, String sendingOrganizationIdentifier, String receivingOrganizationIdentifier)
+	{
+		List<ListResource.ListEntryComponent> binaryIds = new ArrayList<>();
+
+		if (resource instanceof ListResource list)
+			binaryIds.addAll(decryptValidateAndInsertDataStreams(list, privateKey, sendingOrganizationIdentifier,
+					receivingOrganizationIdentifier));
+		else if (resource instanceof Binary binary)
+			binaryIds.add(decryptValidateAndInsertDataResource(binary, privateKey, sendingOrganizationIdentifier,
+					receivingOrganizationIdentifier));
+		else
+			throw new RuntimeException(
+					"Expected resource type Binary or List, got '" + resource.getResourceType().name() + "'");
+
+		return binaryIds;
+	}
+
+	private List<ListResource.ListEntryComponent> decryptValidateAndInsertDataStreams(ListResource list,
+			PrivateKey privateKey, String sendingOrganizationIdentifier, String receivingOrganizationIdentifier)
+	{
+		return list.getEntry().stream().filter(ListResource.ListEntryComponent::hasItem)
+				.filter(e -> e.hasExtension(ConstantsDataSharing.EXTENSION_LIST_ENTRY_MIMETYPE))
+				.map(e -> decryptValidateAndInsertDataStream(e, privateKey, sendingOrganizationIdentifier,
+						receivingOrganizationIdentifier))
+				.toList();
+	}
+
+	private ListResource.ListEntryComponent decryptValidateAndInsertDataStream(ListResource.ListEntryComponent item,
+			PrivateKey privateKey, String sendingOrganizationIdentifier, String receivingOrganizationIdentifier)
+	{
+		String mimeType = getMimeType(item);
+		InputStream inputStream = decryptDataStream(item, privateKey, sendingOrganizationIdentifier,
+				receivingOrganizationIdentifier);
+		validateDataStream(inputStream, mimeType);
+		return insertDataStream(inputStream, mimeType);
+	}
+
+	private ListResource.ListEntryComponent decryptValidateAndInsertDataResource(Binary resource, PrivateKey privateKey,
+			String sendingOrganizationIdentifier, String receivingOrganizationIdentifier)
+	{
+		Binary binary = decryptDataResource(resource, privateKey, sendingOrganizationIdentifier,
+				receivingOrganizationIdentifier);
+		validateDataResource(binary);
+		return insertDataResource(binary);
+	}
+
+	private InputStream decryptDataStream(ListResource.ListEntryComponent listEntry, PrivateKey privateKey,
+			String sendingOrganizationIdentifier, String receivingOrganizationIdentifier)
+	{
+		try
+		{
+			IdType url = (IdType) listEntry.getItem().getReferenceElement();
+
+			InputStream inputStream = api.getFhirWebserviceClientProvider().getWebserviceClient(url.getBaseUrl())
+					.withRetry(ConstantsBase.DSF_CLIENT_RETRY_6_TIMES, ConstantsBase.DSF_CLIENT_RETRY_INTERVAL_5MIN)
+					.readBinary(url.getIdPart(), MediaType.valueOf(MediaType.APPLICATION_OCTET_STREAM));
+
+			// inputStream = RsaAesGcmUtil.decrypt(privateKey, inputStream, sendingOrganizationIdentifier,
+			// receivingOrganizationIdentifier);
+
```

Comment:
remove?

Reply:
It should have been uncommented -> fixed in 5675baa